### PR TITLE
[SPMD] Support SPMDShardToFullShape

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request enables SPMDShardToFullShape. The trickiest part is how to get the full shape, and here is a couple of options:
- [ ] Bookkeeping the shape full shape that enters SPMDFullToShardShape. This is not selected given the output could be created on the fly.
- [ ] Constructing the full shape from the local shard and the sharding spec. This is not selected given there is no way to deal with the padding. We can't examine the data during the tracing time.
- [x] Let users pass the full shape in. This is selected because it's just the most sounded path.

Tes Plan:
PJRT_DEVICE=TPU python test/spmd/test_xla_sharding.py -v -k test_manual_sharding_e2e -k test_spmd_shard_to_full_shape

### CODE REVIEW

File: test/spmd/test_xla_sharding.py [hunk 1]
```diff
@@ -1151,6 +1151,55 @@ def test_spmd_full_to_shard_shape(self):
...
+      x = torch_xla._XLAC._spmd_shard_to_full_shape(
+          xt.global_tensor,
+          torch_xla._XLAC.OpSharding([], [], [], xs.ShardingType.REPLICATED),
+          x.shape, x.dtype)
+
+    xs.clear_sharding(xt)
+    xt = xs._mark_manual_sharding(xt)
+    xx = torch_xla._XLAC._spmd_shard_to_full_shape(
+        xt.global_tensor,
+        torch_xla._XLAC.OpSharding([], [], [], xs.ShardingType.REPLICATED),
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:1175]: So it uses the passed sharding type when returns back to SPMD?

File: test/spmd/test_xla_sharding.py [hunk 2]
```diff
@@ -1151,6 +1151,55 @@ def test_spmd_full_to_shard_shape(self):
...
     self.assertEqual(torch_xla._XLAC._get_xla_sharding_spec(xx), "{manual}")
 
+  def test_spmd_shard_to_full_shape(self):
+    x = torch.zeros(8, 8).to(xm.xla_device())
+    x += 1
+    # No sharding spec attached.
+    with self.assertRaises(RuntimeError):
+      x = torch_xla._XLAC._spmd_shard_to_full_shape(
+          x, torch_xla._XLAC.OpSharding([], [], [], xs.ShardingType.REPLICATED),
+          x.shape, x.dtype)
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:1161]: so we've decided to pass the original full shape as argument, right?

**Review by yeounoh — APPROVED** [review]
LGTM, left some questions -- thanks @alanwaketan ❤️

File: test/spmd/test_xla_sharding.py (hunk 2 above)
Comment (alanwaketan) [reply test/spmd/test_xla_sharding.py:1161]: At least I have decided haha, see the PR descriptions for the reasons behind.

File: test/spmd/test_xla_sharding.py (hunk 1 above)
Comment (alanwaketan) [reply test/spmd/test_xla_sharding.py:1175]: Yup!

**Comment by alanwaketan** [comment]
Thanks @yeounoh for the quick review!
